### DISCUSSION TITLE
What is the status of msvc with scons 4.5.2

### DISCUSSION THREAD

**Discussion by peterphonic** [SCons/scons#4412] [General]
Hello,

We're currently using scons 4.0.1 in conjunction with Visual Studio 2017. To create the environment, we use:
``` 
Environment(MSVC_VERSION=14.1)
```

We plan to upgrade both our compiler to 2022 and scons to version 4.5.2. From the Issues pages, I gathered that post-2017, we can't rely on `vswhere` to determine the installed version. What's the status of 4.5.2 in this regard? Is the following approach suitable for setting up my environment?
```
Environment(MSVC_VERSION=14.3)
```

Additionally, I'm trying to wrap my head around the toolset versioning in relation to Visual Studio versions. Specifically, does 14.3 correspond to Visual Studio 2022?

Thank you in advance.

**Comment by mwichmann** [comment]
there's a table in the manpage now to try to sort this - look at MSVC_VERSION. Sorry can't paste link from this device.

**Reply by peterphonic** [reply]
got it thank

**Comment by bdbaddog** [comment]
> We plan to upgrade both our compiler to 2022 and scons to version 4.5.2. From the Issues pages, I gathered that post-2017, we can't rely on `vswhere` to determine the installed version. What's the status of 4.5.2 in this regard? 

Not sure where you found that information.
vswhere is the definitive way to find post 2017 installs.

If you're doing a vanilla MSVC 2022 install with 4.5.2 or above, I wouldn't expect you to have any issues.
If you're doing something complicated or installing only a preview version of the compiler, you may run into issues.

**Reply by peterphonic** [reply]
Sorry about the incorrect information, I thought I read that starting with vs2017, vswhere was not there any longer or broken, because it was possible to install older toolset.  Like if we install vs2022 but with v141 toolset.

**Reply by peterphonic** [reply]
Always read the doc they said...

So, if I understand well, there are more options to pass to the environment to get it right. Something like that 
`env = Environment(MSVC_VERSION='14.3', MSVC_SCRIPT_ARGS='10.0.20348.0 -vcvars_ver=14.29.30133')`

So, it is now possible to have a more recent version of visual studio and using an older toolset.

**Comment by mwichmann** [comment]
No, we heavily depend on vswhere to find versions of msvc which it supports.  The ability to support not just multiple concurrent msvc installs but also multiple toolsets under a single msvc install undoubtedly complicated things, but it should be pretty well handled now, at the cost of greater selection complexity than in old SCons versions.

Edit - that is, greater complexity in selecting things that are "not default".  The default case is still simple.

**Reply by peterphonic** [reply]
So, if I have vs2022 and I want to use default toolset from 2022, this should be enough : 
`env = Environment(MSVC_VERSION='14.3')`

The complexity arises if I have vs2022 with older toolset installed and I want to compile against older toolset?

**Comment by mwichmann** [comment]
Yes. Indeed, you shouldn't have to specify the version at all to get the newest compiler/toolset.

**Comment by jcbrill** [comment]
Some recommendations and examples of using older msvc toolsets with recent visual studio installations are given below.

> So, it is now possible to have a more recent version of visual studio and using an older toolset.

Yes, this is possible.

> So, if I understand well, there are more options to pass to the environment to get it right. Something like that
env = Environment(MSVC_VERSION='14.3', MSVC_SCRIPT_ARGS='10.0.20348.0 -vcvars_ver=14.29.30133')

Yes, this is possible.

However, the SDK version (`10.0.20348`) and the msvc toolset version (`14.29.30133`) can be specified directly using scons construction variables.  Whenever possible, it is recommended to use the scons construction variables rather than passing msvc batch file arguments directly.  The scons construction variables are validated and may result in more meaningful error messages.

Msvc batch file arguments with direct scons construction variable implementations include:

| Argument | Construction Variable | Script Argument Equivalent |
| :- | :- | :- |
| UWP | MSVC_UWP_APP=True | MSVC_SCRIPT_ARGS='store' |
| SDK | MSVC_SDK_VERSION='10.0.20348.0' | MSVC_SCRIPT_ARGS='10.0.20348.0' |
| Toolset | MSVC_TOOLSET_VERSION='14.31.31103' | MSVC_SCRIPT_ARGS='-vcvars_ver=14.31.31103' |
| Spectre | MSVC_SPECTRE_LIBS=True | MSVC_SCRIPT_ARGS='-vcvars_spectre_libs=spectre' |

The table above is taken from the internal visual studio documentation and notes:
https://github.com/SCons/scons/tree/master/SCons/Tool/MSCommon#visual-studio-implementation-notes


Msvc batch file errors due to user-specified arguments (`MSVC_SCRIPT_ARGS`) may result in build errors than can be difficult to diagnose.  In some cases, errors may be detected.  In other cases, arguments may simply be ignored by the msvc batch files.

Use of the `MSVC_SCRIPT_ARGS` construction variable is generally discouraged and is only recommended in the following situations:
* msvc batch file arguments are introduced in future visual studio versions that are not yet supported directly, or
* there is a bug in the argument construction variable implementation, validation, and/or error detection.


Your example above can be specified using construction variables as:
```
env = Environment(MSVC_VERSION='14.3', MSVC_SDK_VERSION='10.0.20348.0', MSVC_TOOLSET_VERSION='14.29.30133')
```

Unlike the SDK version which requires specification of the full version number, the msvc toolset version can be specified with a partial version number.  The msvc batch files evaluate the toolset version number specifications in descending order and the first match is selected.

For example, the following are equivalent when `14.29.30133` is the latest v142 toolset version installed in VS2022:
```
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.29.30133')
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.29.30')
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.29')
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.2')
```

The minimum msvc toolset version specifications for the latest version of each major build tools version installed in VS2022 are:
```
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.3') # latest v143 toolset
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.2') # latest v142 toolset
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.1') # latest v141 toolset
env = Environment(MSVC_VERSION='14.3', MSVC_TOOLSET_VERSION='14.0') # latest v140 toolset
```

At present, additional supported msvc arguments are not taken into account when selecting the visual studio instance.  This means that a visual studio installation is selected and then the arguments are applied to the selected msvc instance which may result in errors depending on installed features.

The default visual studio instance (i.e., the most recently released visual studio) is used when `MSVC_VERSION` is not specified.  The following example is a request for the latest v142 toolset installed in the default visual studio:
```
env = Environment(MSVC_TOOLSET_VERSION='14.2') # latest v142 toolset in default VS
```

The above will be successful for:
* VS2022 with the v142 build tools installed
* VS2019

The above will fail for:
* VS2022 without the v142 build tools installed
* VS2017 and earlier
